sql/logictest: TestLogic failed

sql/logictest.TestLogic [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3906343&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3906343&tab=artifacts#/) on master @ [5cd5be83218475ef0cc3c62a856b8d9d22cf142b](https://github.com/cockroachdb/cockroach/commits/5cd5be83218475ef0cc3c62a856b8d9d22cf142b):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2fc2332]

```
Stack: 

```
goroutine 8234028 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc00a12cb40, {0x51cc818, 0xc00c131c90})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:254 +0xaa
panic({0x3849d20, 0x734c7f0})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal.func1()
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:674 +0x192
panic({0x3849d20, 0x734c7f0})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*syntheticDescriptors).set(0xc00a0b8f38, {0x0, 0x1, 0xc0138d45f0})
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/synthetic_descriptors.go:39 +0x5a
github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).SetSyntheticDescriptors(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/collection.go:384
github.com/cockroachdb/cockroach/pkg/sql.(*Server).newConnExecutorWithTxn(0x0, {0x51cc818, 0xc00df7ed90}, 0x203000, 0xc0003f7a10, {0x5310b20, 0xc02668bd00}, 0x0, {{0xc01439e7e0, 0xc0041fdda0}, ...}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:931 +0x2e5
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx(0xc031222a20, {0x51cc818, 0xc00df7ed90}, 0xc005902a50, {0x51580d8, 0xc02668bbc0}, 0xc0119c3b80, 0x0, 0xc003a34940, 0xc0003f79e0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:187 +0x3ef
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal(0xc031222a20, {0x51b0c78, 0xc01fada540}, {0x3d562b3, 0xe}, 0xc02668bbc0, 0xc00b5dc240, {{{0x3d3cbda, 0x4}}, {0x0, ...}, ...}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:732 +0x725
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).queryInternalBuffered(0x0, {0x51b0c78, 0xc01fada540}, {0x3d562b3, 0xe}, 0x0, {{{0x3d3cbda, 0x4}}, {0x0, 0x0}, ...}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:427 +0x19b
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowExWithCols(0xc00d2cad50, {0x51b0c78, 0xc01fada540}, {0x3d562b3, 0x425965}, 0x8, {{{0x3d3cbda, 0x4}}, {0x0, 0x0}, ...}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:485 +0xbb
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowEx(0x0, {0x51b0c78, 0xc01fada540}, {0x3d562b3, 0x0}, 0x0, {{{0x3d3cbda, 0x4}}, {0x0, 0x0}, ...}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:471 +0xa5
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).load.func1({0x51b0c78, 0xc01fada540}, 0x203002)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:776 +0x33c
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).runInTxn(0x160, {0x51b0c78, 0xc01fada540}, 0x160, 0xc00074a000)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:737 +0x37
github.com/cockroachdb/cockroach/pkg/jobs.(*Job).load(0xc0158e6c60, {0x51b0c78, 0xc01fada540}, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/jobs.go:766 +0x9c
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).LoadJobWithTxn(0xc020ee9440, {0x51b0c78, 0xc01fada540}, 0x9fa47fcb8300001, 0x1c28f57)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:722 +0x7f
github.com/cockroachdb/cockroach/pkg/sql/gcjob.initDetailsAndProgress.func1({0x51b0c78, 0xc01fada540}, 0xc00034c190)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job_utils.go:69 +0x74
github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1({0x51b0c78, 0xc01fada540}, 0xc00ba37680)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:849 +0x27
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc005902a50, {0x51b0c78, 0xc01fada540}, 0xc00123d3b0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:961 +0xbf
github.com/cockroachdb/cockroach/pkg/kv.runTxn({0x51b0c78, 0xc01fada540}, 0xc01450cd80, 0x40)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:848 +0x5a
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0x0, {0x51b0c78, 0xc01fada540}, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:830 +0x85
github.com/cockroachdb/cockroach/pkg/sql/gcjob.initDetailsAndProgress({0x51b0c78, 0xc01fada540}, 0xc0043d0400, 0x9fa47fcb8300001)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job_utils.go:67 +0xd5
github.com/cockroachdb/cockroach/pkg/sql/gcjob.schemaChangeGCResumer.Resume({0x9fa47fcb8300001}, {0x51b0c78, 0xc01fada540}, {0x3b75c00, 0xc000010858})
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:181 +0x151
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2(0x51cc818, {0x5158920, 0xc02dbde478}, {0x51b0c78, 0xc01fada540}, {0x3b75c00, 0xc000010858}, 0xc00123d948)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1209 +0xba
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc020ee9440, {0x51cc818, 0xc00034c190}, {0x3b75c00, 0xc000010858}, {0x5158920, 0xc02dbde478}, 0xc018f77ce0, {0xc003a34290, 0x7}, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1210 +0x6b7
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc020ee9440, {0x51b0bd0, 0xc0042c0ac0}, {0x5158920, 0xc02dbde478}, 0xc018f77ce0, {0xc003a34290, 0x7}, {0xc0184a22a0, 0x16})
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:414 +0x54f
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1({0x51cc818, 0xc00f056790})
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:336 +0xfa
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:491 +0x12a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:481 +0x445
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestLogic/fakedist-metadata/schema_change_in_txn/check_on_new_table
```
</p>
</details>
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-parallel=2
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #63262 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-20.2]
- #61657 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-21.1]
- #58269 sql/logictest: TestLogic failed [C-test-failure O-robot T-sql-schema branch-release-20.1]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/logictest.TestLogic [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3937389&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3937389&tab=artifacts#/) on master @ [3b4e180a23f5121e0d4106eb3dc5f61ebc314188](https://github.com/cockroachdb/cockroach/commits/3b4e180a23f5121e0d4106eb3dc5f61ebc314188):


```
=== RUN   TestLogic
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic3375069071
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestLogic
    logic.go:3587: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic3375069071
--- FAIL: TestLogic (2619.93s)
=== RUN   TestLogic/fakedist/postgresjoin
=== PAUSE TestLogic/fakedist/postgresjoin
=== CONT  TestLogic/fakedist/postgresjoin
=== CONT  TestLogic/fakedist/postgresjoin
    logic.go:2438: 
         pq: internal error: unexpected error from the vectorized engine: context canceled
--- done: testdata/logic_test/postgresjoin with config fakedist: 40 tests, 1 failures
    logic.go:3353: 
        testdata/logic_test/postgresjoin:1618: error while processing
    logic.go:3353: testdata/logic_test/postgresjoin:1618: too many errors encountered, skipping the rest of the input
        --- FAIL: TestLogic/fakedist/postgresjoin (0.74s)
=== RUN   TestLogic/fakedist
    --- FAIL: TestLogic/fakedist (0.01s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-parallel=2
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #63262 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-20.2]
- #61657 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-21.1]
- #58269 sql/logictest: TestLogic failed [C-test-failure O-robot T-sql-schema branch-release-20.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/logictest.TestLogic [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3940618&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3940618&tab=artifacts#/) on master @ [ce26e48fcc51c66b8e32502cb989cd7cb9d4af66](https://github.com/cockroachdb/cockroach/commits/ce26e48fcc51c66b8e32502cb989cd7cb9d4af66):


```
=== RUN   TestLogic
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic2447271491
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestLogic
    logic.go:3587: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic2447271491
--- FAIL: TestLogic (1271.01s)
=== RUN   TestLogic/5node/distsql_union
=== PAUSE TestLogic/5node/distsql_union
=== CONT  TestLogic/5node/distsql_union
        --- FAIL: TestLogic/5node/distsql_union (3.99s)
=== RUN   TestLogic/5node
    --- FAIL: TestLogic/5node (0.00s)
=== RUN   TestLogic/5node/distsql_union/Intersect
=== CONT  TestLogic/5node/distsql_union/Intersect
    logic.go:2438: 
         pq: inbox communication error: rpc error: code = Canceled desc = context canceled
    logic.go:1984: 
         testdata/logic_test/distsql_union:295: too many errors encountered, skipping the rest of the input
--- done: testdata/logic_test/distsql_union with config 5node: 30 tests, 2 failures
            --- FAIL: TestLogic/5node/distsql_union/Intersect (0.60s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #63262 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-20.2]
- #61657 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-21.1]
- #58269 sql/logictest: TestLogic failed [C-test-failure O-robot T-sql-schema branch-release-20.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/alter_table.go
+++ b/pkg/sql/alter_table.go
@@ -809,7 +809,8 @@ func (n *alterTableNode) startExec(params runParams) error {
 						"constraint %q in the middle of being added, try again later", t.Constraint)
 				}
 				if err := validateCheckInTxn(
-					params.ctx, &params.p.semaCtx, params.ExecCfg().InternalExecutor, params.SessionData(), n.tableDesc, params.EvalContext().Txn, ck.Expr,
+					params.ctx, &params.p.semaCtx, params.ExecCfg().InternalExecutorFactory,
+					params.SessionData(), n.tableDesc, params.EvalContext().Txn, ck.Expr,
 				); err != nil {
 					return err
 				}
@@ -831,8 +832,12 @@ func (n *alterTableNode) startExec(params runParams) error {
 						"constraint %q in the middle of being added, try again later", t.Constraint)
 				}
 				if err := validateFkInTxn(
-					params.ctx, params.p.LeaseMgr(), params.ExecCfg().InternalExecutor,
-					n.tableDesc, params.EvalContext().Txn, name, params.EvalContext().Codec,
+					params.ctx, params.p.LeaseMgr(),
+					params.ExecCfg().InternalExecutorFactory,
+					params.p.SessionData(),
+					n.tableDesc,
+					params.EvalContext().Txn,
+					name, params.EvalContext().Codec,
 				); err != nil {
 					return err
 				}
@@ -855,7 +860,9 @@ func (n *alterTableNode) startExec(params runParams) error {
 							"constraint %q in the middle of being added, try again later", t.Constraint)
 					}
 					if err := validateUniqueWithoutIndexConstraintInTxn(
-						params.ctx, params.ExecCfg().InternalExecutor, n.tableDesc, params.EvalContext().Txn, name,
+						params.ctx, params.ExecCfg().InternalExecutorFactory(
+							params.ctx, params.SessionData(),
+						), n.tableDesc, params.EvalContext().Txn, name,
 					); err != nil {
 						return err
 					}

--- a/pkg/sql/backfill.go
+++ b/pkg/sql/backfill.go
@@ -152,7 +152,7 @@ func (sc *SchemaChanger) makeFixedTimestampRunner(readAsOf hlc.Timestamp) histor
 			ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
 		) error {
 			// We need to re-create the evalCtx since the txn may retry.
-			evalCtx := createSchemaChangeEvalCtx(ctx, sc.execCfg, readAsOf, sc.ieFactory, descriptors)
+			evalCtx := createSchemaChangeEvalCtx(ctx, sc.execCfg, readAsOf, descriptors)
 			return retryable(ctx, txn, &evalCtx)
 		})
 	}
@@ -168,9 +168,7 @@ func (sc *SchemaChanger) makeFixedTimestampInternalExecRunner(
 			ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
 		) error {
 			// We need to re-create the evalCtx since the txn may retry.
-			ie := createSchemaChangeEvalCtx(
-				ctx, sc.execCfg, readAsOf, sc.ieFactory, descriptors,
-			).SchemaChangeInternalExecutor
+			ie := sc.ieFactory(ctx, NewFakeSessionData(sc.execCfg.SV()))
 			return retryable(ctx, txn, ie)
 		})
 	}
@@ -713,21 +711,21 @@ func (sc *SchemaChanger) validateConstraints(
 				defer func() { collection.ReleaseAll(ctx) }()
 				if c.IsCheck() {
 					if err := validateCheckInTxn(
-						ctx, &semaCtx, evalCtx.SchemaChangeInternalExecutor, evalCtx.SessionData(), desc, txn, c.Check().Expr,
+						ctx, &semaCtx, sc.ieFactory, evalCtx.SessionData(), desc, txn, c.Check().Expr,
 					); err != nil {
 						return err
 					}
 				} else if c.IsForeignKey() {
-					if err := validateFkInTxn(ctx, sc.leaseMgr, evalCtx.SchemaChangeInternalExecutor, desc, txn, c.GetName(), evalCtx.Codec); err != nil {
+					if err := validateFkInTxn(ctx, sc.leaseMgr, sc.ieFactory, evalCtx.SessionData(), desc, txn, c.GetName(), evalCtx.Codec); err != nil {
 						return err
 					}
 				} else if c.IsUniqueWithoutIndex() {
-					if err := validateUniqueWithoutIndexConstraintInTxn(ctx, evalCtx.SchemaChangeInternalExecutor, desc, txn, c.GetName()); err != nil {
+					if err := validateUniqueWithoutIndexConstraintInTxn(ctx, sc.ieFactory(ctx, evalCtx.SessionData()), desc, txn, c.GetName()); err != nil {
 						return err
 					}
 				} else if c.IsNotNull() {
 					if err := validateCheckInTxn(
-						ctx, &semaCtx, evalCtx.SchemaChangeInternalExecutor, evalCtx.SessionData(), desc, txn, c.Check().Expr,
+						ctx, &semaCtx, sc.ieFactory, evalCtx.SessionData(), desc, txn, c.Check().Expr,
 					); err != nil {
 						// TODO (lucy): This should distinguish between constraint
 						// validation errors and other types of unexpected errors, and
@@ -991,7 +989,7 @@ func (sc *SchemaChanger) distIndexBackfill(
 		if err != nil {
 			return err
 		}
-		evalCtx = createSchemaChangeEvalCtx(ctx, sc.execCfg, txn.ReadTimestamp(), sc.ieFactory, descriptors)
+		evalCtx = createSchemaChangeEvalCtx(ctx, sc.execCfg, txn.ReadTimestamp(), descriptors)
 		planCtx = sc.distSQLPlanner.NewPlanningCtx(ctx, &evalCtx, nil /* planner */, txn,
 			true /* distribute */)
 		indexBatchSize := indexBackfillBatchSize.Get(&sc.execCfg.Settings.SV)
@@ -1265,7 +1263,7 @@ func (sc *SchemaChanger) distColumnBackfill(
 				return nil
 			}
 			cbw := MetadataCallbackWriter{rowResultWriter: &errOnlyResultWriter{}, fn: metaFn}
-			evalCtx := createSchemaChangeEvalCtx(ctx, sc.execCfg, txn.ReadTimestamp(), sc.ieFactory, descriptors)
+			evalCtx := createSchemaChangeEvalCtx(ctx, sc.execCfg, txn.ReadTimestamp(), descriptors)
 			recv := MakeDistSQLReceiver(
 				ctx,
 				&cbw,
@@ -2090,7 +2088,8 @@ func runSchemaChangesInTxn(
 			check := &c.ConstraintToUpdateDesc().Check
 			if check.Validity == descpb.ConstraintValidity_Validating {
 				if err := validateCheckInTxn(
-					ctx, &planner.semaCtx, planner.ExecCfg().InternalExecutor, planner.SessionData(), tableDesc, planner.txn, check.Expr,
+					ctx, &planner.semaCtx, planner.ExecCfg().InternalExecutorFactory,
+					planner.SessionData(), tableDesc, planner.txn, check.Expr,
 				); err != nil {
 					return err
 				}
@@ -2187,7 +2186,7 @@ func runSchemaChangesInTxn(
 func validateCheckInTxn(
 	ctx context.Context,
 	semaCtx *tree.SemaContext,
-	ie *InternalExecutor,
+	ief sqlutil.SessionBoundInternalExecutorFactory,
 	sessionData *sessiondata.SessionData,
 	tableDesc *tabledesc.Mutable,
 	txn *kv.Txn,
@@ -2197,6 +2196,7 @@ func validateCheckInTxn(
 	if tableDesc.Version > tableDesc.ClusterVersion.Version {
 		syntheticDescs = append(syntheticDescs, tableDesc)
 	}
+	ie := ief(ctx, sessionData)
 	return ie.WithSyntheticDescriptors(syntheticDescs, func() error {
 		return validateCheckExpr(ctx, semaCtx, sessionData, checkExpr, tableDesc, ie, txn)
 	})
@@ -2217,7 +2217,8 @@ func validateCheckInTxn(
 func validateFkInTxn(
 	ctx context.Context,
 	leaseMgr *lease.Manager,
-	ie *InternalExecutor,
+	ief sqlutil.SessionBoundInternalExecutorFactory,
+	sd *sessiondata.SessionData,
 	tableDesc *tabledesc.Mutable,
 	txn *kv.Txn,
 	fkName string,
@@ -2239,7 +2240,7 @@ func validateFkInTxn(
 	if fk == nil {
 		return errors.AssertionFailedf("foreign key %s does not exist", fkName)
 	}
-
+	ie := ief(ctx, sd)
 	return ie.WithSyntheticDescriptors(syntheticDescs, func() error {
 		return validateForeignKey(ctx, tableDesc, fk, ie, txn, codec)
 	})
@@ -2259,7 +2260,7 @@ func validateFkInTxn(
 // reuse an existing kv.Txn safely.
 func validateUniqueWithoutIndexConstraintInTxn(
 	ctx context.Context,
-	ie *InternalExecutor,
+	ie sqlutil.InternalExecutor,
 	tableDesc *tabledesc.Mutable,
 	txn *kv.Txn,
 	constraintName string,

--- a/pkg/sql/check.go
+++ b/pkg/sql/check.go
@@ -44,7 +44,7 @@ func validateCheckExpr(
 	sessionData *sessiondata.SessionData,
 	exprStr string,
 	tableDesc *tabledesc.Mutable,
-	ie *InternalExecutor,
+	ie sqlutil.InternalExecutor,
 	txn *kv.Txn,
 ) error {
 	expr, err := schemaexpr.FormatExprForDisplay(ctx, tableDesc, exprStr, semaCtx, sessionData, tree.FmtParsable)
@@ -237,7 +237,7 @@ func validateForeignKey(
 	ctx context.Context,
 	srcTable *tabledesc.Mutable,
 	fk *descpb.ForeignKeyConstraint,
-	ie *InternalExecutor,
+	ie sqlutil.InternalExecutor,
 	txn *kv.Txn,
 	codec keys.SQLCodec,
 ) error {

--- a/pkg/sql/index_backfiller.go
+++ b/pkg/sql/index_backfiller.go
@@ -133,7 +133,7 @@ func (ib *IndexBackfillPlanner) plan(
 	if err := DescsTxn(ctx, ib.execCfg, func(
 		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
 	) error {
-		evalCtx = createSchemaChangeEvalCtx(ctx, ib.execCfg, nowTimestamp, ib.ieFactory, descriptors)
+		evalCtx = createSchemaChangeEvalCtx(ctx, ib.execCfg, nowTimestamp, descriptors)
 		planCtx = ib.execCfg.DistSQLPlanner.NewPlanningCtx(ctx, &evalCtx, nil /* planner */, txn,
 			true /* distribute */)
 		// TODO(ajwerner): Adopt util.ConstantWithMetamorphicTestRange for the

--- a/pkg/sql/planner.go
+++ b/pkg/sql/planner.go
@@ -101,8 +101,6 @@ type extendedEvalContext struct {
 	indexUsageStats *idxusage.LocalIndexUsageStats
 
 	SchemaChangerState *SchemaChangerState
-
-	SchemaChangeInternalExecutor *InternalExecutor
 }
 
 // copyFromExecCfg copies relevant fields from an ExecutorConfig.

--- a/pkg/sql/schema_changer.go
+++ b/pkg/sql/schema_changer.go
@@ -1941,24 +1941,18 @@ func (sc *SchemaChanger) txn(
 // used in the surrounding SQL session, so session tracing is unable
 // to capture schema change activity.
 func createSchemaChangeEvalCtx(
-	ctx context.Context,
-	execCfg *ExecutorConfig,
-	ts hlc.Timestamp,
-	ieFactory sqlutil.SessionBoundInternalExecutorFactory,
-	descriptors *descs.Collection,
+	ctx context.Context, execCfg *ExecutorConfig, ts hlc.Timestamp, descriptors *descs.Collection,
 ) extendedEvalContext {
 
 	sd := NewFakeSessionData(execCfg.SV())
-	ie := ieFactory(ctx, sd)
 
 	evalCtx := extendedEvalContext{
 		// Make a session tracing object on-the-fly. This is OK
 		// because it sets "enabled: false" and thus none of the
 		// other fields are used.
-		Tracing:                      &SessionTracing{},
-		ExecCfg:                      execCfg,
-		Descs:                        descriptors,
-		SchemaChangeInternalExecutor: ie.(*InternalExecutor),
+		Tracing: &SessionTracing{},
+		ExecCfg: execCfg,
+		Descs:   descriptors,
 		EvalContext: tree.EvalContext{
 			SessionDataStack: sessiondata.NewStack(sd),
 			// TODO(andrei): This is wrong (just like on the main code path on

